### PR TITLE
crypto: thread: serialize concurrent joins

### PR DESCRIPTION
**Pull request by ckalina**
Multiple concurrent joins with a running thread suffer from a race condition that allows concurrent join calls to perform concurrent arch specific join calls, which is UB on POSIX, or to concurrently execute join and terminate calls.

As soon as a thread T1 exists, one of the threads that joins with T1 is selected to perform the join, the remaining ones await completion. Once completed, the remaining calls immediately return. If the join failed, another thread is selected to attempt the join operation.

Forcefully terminating a thread that is in the process of joining another thread is not supported.

Common code from thread_posix and thread_win was refactored to use common wrapper that handles synchronization.

This should fix the timeout encountered while executing threads test (likely due to test_thread_native_multiple_joins).

##### Checklist

- [x] tests are added or updated

### CODE REVIEW

**Comment by t8m** [comment]
Unfortunately this did not seem to fix the MacOSX hang.

**Comment by ckalina** [comment]
The MacOS hang was indeed independent of this and should now hopefully be fixed.

While POSIX threads are cancellable and may be asynchronously cancelled, their cancellation is not guaranteed by the POSIX standard.

test_thread_noreturn, which simulates a long-running possibly unresponsive thread:
```
	THREAD T1	THREAD T2
-------------------------------------------------------------------------------
	LOCK L1
	SPAWN T2
				LOCK L1
```
On MacOS, cancelling such thread only queues cancellation request, but the following pthread_join hangs.

Replace this implementation by an unbounded sequence of sleeps instead. One second sleeps are chosen, but anything smaller goes too. I'd prefer not to use larger values, though, to ensure that it's woken up often enough.

**Review by t8m — APPROVED** [review]

**Comment by t8m** [comment]
This is urgent. The buildbot macosx failure is unrelated.

**Review by hlandau — APPROVED** [review]

**Comment by hlandau** [comment]
Agree urgent

**Comment by t8m** [comment]
I'll wait for the windows buildbot builds to finish and merge afterwards.

**Comment by t8m** [comment]
For some reason the Windows test still fail when running on buildbot. I'll merge this anyway to unblock at least the macosx builds and we'll continue in another PR.

**Comment by t8m** [comment]
Merged to master branch. Thank you. Assuming a PR fixing the Windows buildbot hang will follow.

**Comment by mouse07410** [comment]
I confirm that this PR fixed the problem on MacOS.

**Pull request closed without merging**
